openrouteservice/src/main/java/org/heigit/ors/matrix/TargetGraphBuilder.java

```diff
@@ -0,0 +1,112 @@
+     * The EdgeFilter in use is a downward search edge filter with reverse access acceptance so that in the last phase of the algorithm, the targetGraph can be explored downwards
+     *
+     * @param targets the targets that form the seed for target graph building
+     */
+    public TargetGraphResults prepareTargetGraph(int[] targets, CHGraph chGraph, Graph graph, FlagEncoder encoder, boolean swap, int coreNodeLevel) {
+        PriorityQueue<Integer> localPrioQueue = new PriorityQueue<>(100);
+        ExclusiveDownwardSearchEdgeFilter downwardEdgeFilter = new ExclusiveDownwardSearchEdgeFilter(chGraph, encoder, swap);
+        EdgeExplorer edgeExplorer = swap ? graph.createEdgeExplorer(DefaultEdgeFilter.outEdges(encoder)) : graph.createEdgeExplorer(DefaultEdgeFilter.inEdges(encoder));
+        SubGraph targetGraph = new SubGraph(graph);
+        IntHashSet coreExitPoints = new IntHashSet();
+
+        this.coreNodeLevel = coreNodeLevel;
+        this.chGraph = chGraph;
+        addNodes(targetGraph, localPrioQueue, targets, coreExitPoints);
+
+        while (!localPrioQueue.isEmpty()) {
+            int adjNode = localPrioQueue.poll();
+            EdgeIterator iter = edgeExplorer.setBaseNode(adjNode);
+            downwardEdgeFilter.setBaseNode(adjNode);
+            exploreEntry(targetGraph, localPrioQueue, downwardEdgeFilter, adjNode, iter, coreExitPoints);
+        }
+        TargetGraphResults targetGraphResults = new TargetGraphResults();
+        targetGraphResults.setTargetGraph(targetGraph);
+        targetGraphResults.setCoreExitPoints(coreExitPoints);
+        return targetGraphResults;
+    }
+
+    /**
+     * Explore the target graph and build coreExitPoints
+     * @param localPrioQueue
+     * @param downwardEdgeFilter
+     * @param adjNode
+     * @param iter
+     */
+    private void exploreEntry(SubGraph targetGraph, PriorityQueue<Integer> localPrioQueue, ExclusiveDownwardSearchEdgeFilter downwardEdgeFilter, int adjNode, EdgeIterator iter, IntHashSet coreExitPoints) {
+        while (iter.next()) {
+            if (!downwardEdgeFilter.accept(iter))
+                continue;
+            boolean isNewNode = targetGraph.addEdge(adjNode, iter, true);
+            if (isCoreNode(iter.getAdjNode()) && !isCoreNode(iter.getBaseNode())) {
+                coreExitPoints.add(iter.getAdjNode());
+            } else if(isNewNode) {
+                localPrioQueue.add(iter.getAdjNode());
+            }
+        }
+    }
+
+    /**
+     * Add nodes to target graph and prioQueue for target graph
+     * @param graph
+     * @param prioQueue
+     * @param nodes
+     */
+    private void addNodes(SubGraph graph, PriorityQueue<Integer> prioQueue, int[] nodes, IntHashSet coreExitPoints) {
+        for (int i = 0; i < nodes.length; i++) {
+            int nodeId = nodes[i];
+            if (nodeId >= 0) {
+                if (graph != null)
+                    graph.addEdge(nodeId, null, true);
+                prioQueue.add(nodeId);
```

Comment:
Nodes are added here to the queue regardless of being in core or not. AFAIU the idea is to explore the tree only up to the core, so I'm not sure this is necessary. Actually, I think that in the current implementation `downwardEdgeFilter` might accept a connection between nodes at `coreNodeLevel` and `coreNodeLevel + 1` in the presence of turn restrictions. This situation, however, seems to be handled by `!isCoreNode(iter.getBaseNode())` in L60. It appears to me that this check might be obsolete if core nodes are not added to the queue if first place, what do you think?

Reply:
Done.

Reply:
Awesome, thanks!